arduino/smartcar/smartcar.ino

```diff
@@ -14,18 +18,77 @@ void setup() {
 }
 
 void loop() {
-  car.setSpeed(90);
-  preventCrash();
+  car.setSpeed(50); 
+  preventCrashWithUltrasonic(); 
+  preventCrashWithInfraredFront(); 
+  preventCrashWithInfraredLeft();
+  preventCrahsWithInfraredRight();
+  preventCrashWithInfraredBack();
 }
 
-void preventCrash() {
-  unsigned int fDistance = front.getDistance();
+void preventCrashWithUltrasonic() {
+  unsigned int fDistance = frontUS.getDistance();
   if (fDistance > 0 && fDistance < 200) {
+    Serial.print("Obstacle detected by front ultrasonic sensor");
     car.setSpeed(0);
     delay(500);
-    control.overrideMotorSpeed(100, -100);
+    control.overrideMotorSpeed(50, -100);
     delay(800);
     car.setSpeed(0);
     delay(200);
   }
 }
+
+void preventCrashWithInfraredFront() {
+  unsigned int fDistance = frontIR.getDistance();
+  if (fDistance > 0 && fDistance < 200) {
+    Serial.print("Obstacle detected by front infrared sensor"); 
+    car.setSpeed(0);
+    delay(500);
+    control.overrideMotorSpeed(50, -100);
+    delay(800);
+    car.setSpeed(0);
+    delay(200);
+  }
+}
+
+void preventCrashWithInfraredLeft() {
```

Comment:
What good is this method actually? As far as I can tell, when the car is passing by a tree on its left side that it is absolutely not on a collision course with, it will stop and turn around. I can see the value of detecting an obstacle to the sides, for example to prevent turning in that direction at a particularly sharp angle, where the front sensor might not detect the obstacle in time. But if there is no imminent crash the car should not be stopped.

Reply:
Thresholds were adapted to avoid this behavior.

Reply:
Tweaking the range for detection doesn't change this behavior. The car might still pass by an obstacle on its left side and stop and turn even though there was no danger of collision. If you want to make use of sensor data from the sides of the car I would expect something like calculating a dangerous turn-angle from the distance to the object and preventing it from taking that angle. Otherwise I don't see what it's good for (other than recording the obstacle for the purpose of creating a map maybe?).

Reply:
The US sensor detects obstacles in the range of 200 cm to 4000 cm, while the IR detects obstacles between 0 cm and 400 cm. In order to complement the US sensor, the front IR is now in charged to detect obstacles in the range missed by the US.

Reply:
@alynnka check the distances, they are not correct.
What @FabianSchlemmer means is that if you pass by an obstacle on your right while moving forward, you don't need to make a sharp turn to avoid it since your current trajectory will not lead to a collision.

Reply:
Distances are fixed now: the US sensor detects obstacles in the range of 20 cm to 400 cm, while the IR detects obstacles between 0 cm and 40 cm . The left, right and back IR sensors are not longer used.